src/engine/idTable/CompressedExternalIdTable.h

```diff
@@ -209,49 +211,96 @@ class CompressedExternalIdTableWriter {
+          index_(index),
+          firstBlock_(writer->startOfSingleIdTables_.at(index)),
+          lastBlock_(index + 1 < writer->startOfSingleIdTables_.size()
+                         ? writer->startOfSingleIdTables_.at(index + 1)
+                         : writer->blocksPerColumn_.at(0).size()),
+          blockIdx_(firstBlock_) {
+      ++writer_->numActiveGenerators_;
+    }
+    ~TableBlockGenerator() {
+      if (!hasDecremented_) {
+        --writer_->numActiveGenerators_;
       }
-      fut = std::async(std::launch::async, [this, blockIdx]() {
-        return readBlock<NumCols>(blockIdx);
+    }
+
+    std::optional<IdTableStatic<NumCols>> operator()() {
+      absl::Cleanup cleanup([this] {
+        --writer_->numActiveGenerators_;
+        hasDecremented_ = true;
       });
-      if (table.has_value()) {
-        co_yield table.value();
+      if (isFirstCall_) {
+        isFirstCall_ = false;
+        if (blockIdx_ >= lastBlock_) {
+          return std::nullopt;
+        }
+        fut_ = std::async(std::launch::async, [this]() {
+          return this->writer_->template readBlock<NumCols>(this->blockIdx_++);
+        });
+        AD_CONTRACT_CHECK(
+            fut_.valid(),
+            "Future initialization failed in TableBlockGenerator");
+        auto table = fut_.get();
+        if (blockIdx_ < lastBlock_) {
+          fut_ = std::async(std::launch::async, [this]() {
+            return this->writer_->template readBlock<NumCols>(
+                this->blockIdx_++);
+          });
+          std::move(cleanup).Cancel();
+        }
+        return table;
+      }
+      if (blockIdx_ > lastBlock_) {
+        return std::nullopt;
+      }
+      if (!fut_.valid()) {
+        return std::nullopt;
       }
+      auto table = fut_.get();
+      if (blockIdx_ < lastBlock_) {
+        fut_ = std::async(std::launch::async, [this]() {
+          return this->writer_->template readBlock<NumCols>(this->blockIdx_++);
+        });
+        std::move(cleanup).Cancel();
+      }
+      return table;
     }
-    // Last block.
-    AD_CORRECTNESS_CHECK(fut.valid());
-    auto table = fut.get();
-    co_yield table;
-    --numActiveGenerators_;
-    std::move(cleanup).Cancel();
+  };
+
+  template <size_t NumCols = 0>
+  InputRangeTypeErased<IdTableStatic<NumCols>> makeGeneratorForIdTable(
+      size_t index) {
+    auto generator =
+        std::make_shared<TableBlockGenerator<NumCols>>(this, index);
+    return InputRangeTypeErased<IdTableStatic<NumCols>>{
+        InputRangeFromGetCallable{[g = std::move(generator), this,
+                                   generatorStarted = false]() mutable {
+          if (!generatorStarted) {
+            generatorStarted = true;
+            ++this->numActiveGenerators_;
+          }
+          auto result = (*g)();
+          if (!result.has_value() && !g->hasDecremented_) {
+            g->hasDecremented_ = true;
+            --numActiveGenerators_;
+          }
+          return result;
```

Comment:
Do we really need the handling of the num generators in both classes?
Can't just the `TableBlockGenerator` increate the `numActiveGenerators` in its constructor, 
and decrease it either when there's an error, or when it has yielded the last block or no more active future, 
s.t. this function here is a simple `return InputRangeTypeErased { TableBlockGenerator {...}}` ?